api/src/main/kotlin/com/openlattice/search/Auth0UserSearchFields.kt

```diff
@@ -0,0 +1,20 @@
+package com.openlattice.search
+
+import com.fasterxml.jackson.annotation.JsonProperty
+
+// https://auth0.com/docs/users/user-search/user-search-query-syntax
+data class Auth0UserSearchFields(
+    @JsonProperty val email: String?,
+    @JsonProperty val name: String?
+) {
+    init {
+        // TODO - support multiple fields and construct a valid Lucene query string to pass to Auth0
+        require((email != null && name != null).not()) { "only one of \"email\", \"name\" are allowed" }
+        if (email != null) {
+            require(email.isNotBlank()) { "email cannot be blank" }
+        }
+        if (name != null) {
+            require(name.isNotBlank()) { "name cannot be blank" }
+        }
+    }
```

Comment:
Looks like the use case is to have only one or the other field filled but not both. 
One way I like of doing that is as follows using factory methods:

```
class Auth0UserSearchFields private constructor( val email: String?, val name: Int? ) {
    companion object {
        @JvmStatic
        fun emailSearch( val email ): Auth0UserSearchFields {
            return Auth0UserSearchFields(email, null)
        }
        
        @JvmStatic
        fun nameSearch( val name ): Auth0UserSearchFields {
            return Auth0UserSearchFields(null, name)
        }
    }
}
```

That way the user of this class never has the option of providing an invalid search that has both name and email because the constructor is private

Reply:
I like it, but it's not what I was going for. The query Auth0 accepts is a Lucene query so it's totally possible to do the following:

```
name:test* AND email:test@openlattice.com
```

The issue is basically that I'm not sure where the "Lucene query builder" logic belongs ... is it on the frontend? ... is it on the backend? ... is it a mixture of both? I don't know. Additionally, we can search other Auth0 fields like `given_name`, `family_name, `email_verified`, etc. While we don't have a use case for that at the moment, I thought it would be easier to add support for additional fields in the future if we have a simple data class.

Reply:
Gotcha, yeah that makes sense